File: stack/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []

        brackets = {
            ')' : '(',
            '}' : '{',
            ']' : '['
        }

        for bracket in s:
            if bracket not in brackets:
                stack.append(bracket)
            else:
                if not stack or brackets[bracket] != stack.pop():
                    return False
        return True if not stack else False

File: stack/valid_parentheses.cpp
class Solution {
public:
  bool isValid(string s) {
    std::stack<char> paren_stack;

    std::unordered_map<char, char> brackets = {
        {')', '('}, {'}', '{'}, {']', '['}};

    for (char b : s) {
      if (brackets.count(b) > 0) {
        if (!paren_stack.empty() && paren_stack.top() == brackets[b]) {
          paren_stack.pop();
        } else {
          return false;
        }
      } else {
        paren_stack.push(b);
      }
    }
    return paren_stack.empty();
  }
};
